protocols/api/src/main/java/org/apache/james/protocols/api/ProtocolSession.java

```diff
@@ -44,7 +100,7 @@
      * @param state the {@link State} to which the mapping belongs
      * @return oldValue the value which was stored before for this key or <code>null</code> if non was stored before.
      */
-    Object setAttachment(String key, Object value, State state);
+    <T> Optional<T> setAttachment(AttachmentKey<T> key, T value, State state);
```

Comment:
SMTP hooks are an extension mechanism in James people are relying on.

We need to emulate the previous API (with default methoods?) on top of the new proposed one, just like we did for mailet.

Reply:
we did not when we have changed the `User`, it's the same kind of modification.

Reply:
> we did not when we have changed the User, it's the same kind of modification.

I'm surprised that I did not complain about it.

Please then evaluate the impact and document the change in the upgrade instructions for SMTP hooks developpers...